src/compiler/crystal/semantic/simple_rescues.cr

```diff
@@ -0,0 +1,100 @@
+require "set"
+require "../program"
+require "../syntax/transformer"
+
+module Crystal
+  class SimpleRescues < Transformer
+    @i = 0
+
+    def transform(node : ExceptionHandler)
+      ensure_body = node.ensure
+      rescues = node.rescues
+
+      if ensure_body
+        # if there is an ensure wrap it in another begin/rescue
+        # add the original ensure at the end of the outer rescue and
+        # also following the outer rescue
+        node.ensure = nil
+        var_name = next_var
+
+        ensure_body = ensure_body.transform(self)
+
+        body = if rescues || node.else
+                 node
+               else
+                 node.body
+               end
+        new_handler = ExceptionHandler.new(
+          body.transform(self),
+          [Rescue.new(Expressions.from([
+            ensure_body,
+            Call.global("raise", Var.new(var_name)).at(node),
+          ] of ASTNode), nil, var_name)])
+        tap_block = Block.new(@args = [] of Var, ensure_body.clone)
+        return Expressions.from [Call.new(new_handler, "tap", [] of ASTNode, tap_block)] of ASTNode
```

Comment:
We shouldn't use `tap` for this, creating new dependencies in the stdlib. I'd rather you just implemented this as
```cr
___tempvar = begin ... end
ensure_body
___tempvar
```

Reply:
Actually, the code was like that, I had to use tap to fix another issue. Please, read [this comment](https://github.com/crystal-lang/crystal/pull/3582#issuecomment-305200663). Considering the  purpose of SimpleRescues, can we leave it as it is?

Reply:
but the code I put above *is* equivalent to the `tap` call, since it captures `__tempvar` and ensures it's the last item of the `Expressions` so that it has the correct type.

Reply:
Yeah, you are right, I just misunderstood the code.